cli/tools/doc.rs

```diff
@@ -6,74 +6,62 @@ use crate::file_fetcher::File;
 use crate::flags::Flags;
 use crate::get_types;
 use crate::media_type::MediaType;
-use crate::module_graph;
 use crate::program_state::ProgramState;
-use crate::specifier_handler::FetchHandler;
 use crate::write_json_to_stdout;
 use crate::write_to_stdout_ignore_sigpipe;
 use deno_core::error::AnyError;
+use deno_core::futures::future;
 use deno_core::futures::future::FutureExt;
-use deno_core::futures::Future;
-use deno_core::parking_lot::Mutex;
 use deno_core::resolve_url_or_path;
 use deno_doc as doc;
-use deno_doc::parser::DocFileLoader;
+use deno_graph::create_graph;
+use deno_graph::source::LoadFuture;
+use deno_graph::source::LoadResponse;
+use deno_graph::source::Loader;
+use deno_graph::ModuleSpecifier;
 use deno_runtime::permissions::Permissions;
 use std::path::PathBuf;
-use std::pin::Pin;
 use std::sync::Arc;
-use swc_ecmascript::parser::Syntax;
 
-type DocResult = Result<(Syntax, String), doc::DocError>;
-
-/// When parsing lib.deno.d.ts, only `DocParser::parse_source` is used,
-/// which never even references the loader, so this is just a stub for that scenario.
-///
-/// TODO(Liamolucko): Refactor `deno_doc` so this isn't necessary.
 struct StubDocLoader;
 
-impl DocFileLoader for StubDocLoader {
-  fn resolve(
-    &self,
-    _specifier: &str,
-    _referrer: &str,
-  ) -> Result<String, doc::DocError> {
-    unreachable!()
+impl Loader for StubDocLoader {
+  fn load(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    _is_dynamic: bool,
+  ) -> LoadFuture {
+    Box::pin(future::ready((specifier.clone(), Ok(None))))
   }
+}
 
-  fn load_source_code(
-    &self,
-    _specifier: &str,
-  ) -> Pin<Box<dyn Future<Output = DocResult>>> {
-    unreachable!()
-  }
+struct DocLoader {
+  program_state: Arc<ProgramState>,
 }
 
-impl DocFileLoader for module_graph::Graph {
-  fn resolve(
-    &self,
-    specifier: &str,
-    referrer: &str,
-  ) -> Result<String, doc::DocError> {
-    let referrer =
-      resolve_url_or_path(referrer).expect("Expected valid specifier");
-    match self.resolve(specifier, &referrer, true) {
-      Ok(specifier) => Ok(specifier.to_string()),
-      Err(e) => Err(doc::DocError::Resolve(e.to_string())),
+impl Loader for DocLoader {
+  fn load(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    _is_dynamic: bool,
+  ) -> LoadFuture {
+    let specifier = specifier.clone();
+    let program_state = self.program_state.clone();
+    async move {
+      let result = program_state
+        .file_fetcher
+        .fetch(&specifier, &mut Permissions::allow_all())
+        .await
+        .map(|file| {
+          Some(LoadResponse {
+            specifier: specifier.clone(),
+            maybe_headers: None,
+            content: Arc::new(file.source),
+          })
+        });
```

Comment:
So this is wrong - firstly it completely skips "headers"; secondly it doesn't apply import map. 

Not sure if I should embed import map logic here or should I use `maybe_resolver` field on the `create_graph`: https://docs.rs/deno_graph/0.3.0/deno_graph/fn.create_graph.html

Reply:
Okay, so I added headers handling - I'm forwarding them from `FileFetcher` and `File` struct. However I'm still facing problem with import map (so the question above is still valid).

There are two tests that are "almost passing":
```
---- integration::doc::deno_doc_types_header_direct stdout ----
deno_exe path /Users/biwanczuk/dev/deno/target/debug/deno
test_server starting...
deno_exe args doc --reload http://127.0.0.1:4545/xTypeScriptTypes.js
deno_exe testdata path "/Users/biwanczuk/dev/deno/cli/tests/testdata"
output path /Users/biwanczuk/dev/deno/cli/tests/testdata/doc/types_header.out
OUTPUT
Download http://127.0.0.1:4545/xTypeScriptTypes.js
Download http://127.0.0.1:4545/xTypeScriptTypes.d.ts
Defined in http://127.0.0.1:4545/xTypeScriptTypes.js:1:0

const foo: "foo"


OUTPUT
EXPECTED
Download http://127.0.0.1:4545/xTypeScriptTypes.js
Download http://127.0.0.1:4545/xTypeScriptTypes.d.ts
Defined in http://127.0.0.1:4545/xTypeScriptTypes.d.ts:1:0

const foo: "foo"


EXPECTED
thread 'integration::doc::deno_doc_types_header_direct' panicked at 'pattern match failed', test_util/src/lib.rs:1488:7

---- integration::doc::deno_doc_types_header stdout ----
deno_exe path /Users/biwanczuk/dev/deno/target/debug/deno
deno_exe args doc --reload doc/types_header.ts
deno_exe testdata path "/Users/biwanczuk/dev/deno/cli/tests/testdata"
output path /Users/biwanczuk/dev/deno/cli/tests/testdata/doc/types_header.out
OUTPUT
Download http://127.0.0.1:4545/xTypeScriptTypes.js
Download http://127.0.0.1:4545/xTypeScriptTypes.d.ts
Defined in http://127.0.0.1:4545/xTypeScriptTypes.js:1:0

const foo: "foo"


OUTPUT
EXPECTED
Download http://127.0.0.1:4545/xTypeScriptTypes.js
Download http://127.0.0.1:4545/xTypeScriptTypes.d.ts
Defined in http://127.0.0.1:4545/xTypeScriptTypes.d.ts:1:0

const foo: "foo"


EXPECTED
```

As can be seen the actual output shows that symbols are defined in `.d.ts` files while we expect them to be defined in `.js` files. To be honest I'm not sure what should happen here. CC @kitsonk please advise.

Reply:
Ok, I have take a look at it.  There is some existing evil hackery in there that should have never have been in there IMO:

```rust
    // If the root module has external types, the module graph won't redirect it,
    // so instead create a dummy file which exports everything from the actual file being documented.
    let root_specifier = resolve_url_or_path("./$deno$doc.ts").unwrap();
```

We should have never done that.  What we need to do in `deno_doc` is to "prefer types" when resolving modules out of the graph and checking for the types definition for a file.  Originally the module graph in CLI didn't model this properly, and we should have fixed it far earlier versus just working around it.

I will try to raise the PR for `deno_doc` now, should be pretty straight forward and hopefully the solution will "just work" downstream.

Reply:
Ok, I raised https://github.com/denoland/deno_doc/pull/146 which should fix the the problem that was being worked around, so the "magic" doc module should be removed and doc'ed specifier should just be passed, and `deno_graph` and `deno_doc` will properly take care of the rest, and it should cause the right specifiers to be printed out I think.